essay/swar/latency.md

```diff
@@ -149,32 +173,45 @@ multiply128_to128_bits(unsigned __int128, unsigned __int128):           # @multi
+
+Fundamental physics is impeding further miniaturization of transistors, I speculate that, among other things, the excution units dedicated to SIMD and "scalar" or General Purpose operations would have to be physically separated in the chip, moving data from one side to the other would be ever more expensive.  Actually, In x86_64 similar penalties existed such as having the result of a floating point operation being an input to an integer operation, or viceversa.  I used to do this: I would have 64 bit or 32 bit lanes of integer processing, and to retrieve the most significant bit, the one that I cared about, I would use `MOVMSKPD` and `MOVMSKPS`, instructions that would extract the sign bit, the MSB of the corresponding "D" double-precision floating point (the 64-bit MSB) or the "S" single precision.  I had to use the "floating point" gathering of mask bits because there was no integer equivalent for 32 or 64 bit lane size.  The only "integer" operation is `movmskb` (with either/both `v` and `p` prefixes) that gathers the MSB of each byte... for further aggravation, x86_64, did not have 16-bit lane size gathering of MSB...
+
+ARM Neon is even worse: you don't have a cheap way to process the results of a SIMD boolean test.  After careful analysis for why our portable, 64-bit-bound implementaiton of `strlen` is so close in performance to the "proper" SIMD using ARM Neon, that processes 128 bits of input, I conclude that it is that the translation of the SIMD comparison result to "scalar", which we need to count the trailing bits and thus identify the first null, is about as expensive as never using Neon but doing all in "pure" SWAR, emulating SIMD with integers.  The doubling of the width, consuming 16 bytes at a time instead of 8, does not significantly reduce latencies or increase throughput, because it still requires some heavy lifting in the "scalar" side.  By the way, there is a universe of "horizontal" operations that can be emulated with multiplication, for example, summing the values in all lanes:
+```
+{     A    |   B    |    C    |    D   }:  we want A+B+C+D.
+    Let's say we have *scalar* multiplication by very large integers and we can "cast" all lanes into a single scalar:
+    Then, we can craft a number of
+{     1    |    1   |    1     |    1  }: a scalar that encodes the number 1 every N bits where N is the lane width.
+Let's use the typical multiplication algorithm:
+     A          B        C         D +
+     B          C        D         0 +
+     C          D        0         0 +
+     D          0        0         0 =
+A+B+C+D       B+C+D     C+D        D
+The summation is in the most significant lane.
+```
+The issue with emulating missing SIMD horizontal operations is that you have to combine simple operations that diminish the gain factor of parallelism and require sophisticated mathematics, or what other people would call tricks.
```

Comment:
I'd throw the "or what other people would call tricks" in parenthesis or just drop it altogether.